File: android/apolloui/src/main/java/io/muun/apollo/presentation/ui/fragments/recommended_fee/RecommendedFeePresenter.java
package io.muun.apollo.presentation.ui.fragments.recommended_fee;

import io.muun.apollo.domain.model.WithPaymentContext;
import io.muun.apollo.domain.selector.CurrencyDisplayModeSelector;
import io.muun.apollo.presentation.analytics.AnalyticsEvent;
import io.muun.apollo.presentation.analytics.AnalyticsEvent.S_MORE_INFO_TYPE;
import io.muun.apollo.presentation.ui.base.SingleFragmentPresenter;
import io.muun.apollo.presentation.ui.base.di.PerFragment;

import android.os.Bundle;
import androidx.annotation.Nullable;

import javax.inject.Inject;

@PerFragment
public class RecommendedFeePresenter
        extends SingleFragmentPresenter<RecommendedFeeView, RecommendedFeeParentPresenter>
        implements WithPaymentContext {

    private final CurrencyDisplayModeSelector currencyDisplayModeSel;

    /**
     * Creates a presenter.
     */
    @Inject
    public RecommendedFeePresenter(CurrencyDisplayModeSelector currencyDisplayModeSel) {
        this.currencyDisplayModeSel = currencyDisplayModeSel;
    }

    @Override
    public void setUp(Bundle arguments) {
        super.setUp(arguments);

        view.setCurrencyDisplayMode(currencyDisplayModeSel.get());

        view.setPaymentContext(
                getPaymentContext(),
                getParentPresenter().getPaymentRequest()
        );
    }

    public void confirmFee(double selectedFeeRate) {
        getParentPresenter().confirmFee(selectedFeeRate);
    }

    public void editFeeManually() {
        getParentPresenter().editFeeManually();
    }

    @Nullable
    @Override
    protected AnalyticsEvent getEntryEvent() {
        return new AnalyticsEvent.S_SELECT_FEE();
    }

    public void reportShowSelectFeeInfo() {
        analytics.report(new AnalyticsEvent.S_MORE_INFO(S_MORE_INFO_TYPE.SELECT_FEE));
    }
}

File: android/apolloui/src/main/java/io/muun/apollo/presentation/ui/fragments/recommended_fee/RecommendedFeePresenter.kt
package io.muun.apollo.presentation.ui.fragments.recommended_fee

import android.os.Bundle
import io.muun.apollo.domain.analytics.AnalyticsEvent.S_MORE_INFO
import io.muun.apollo.domain.analytics.AnalyticsEvent.S_MORE_INFO_TYPE
import io.muun.apollo.domain.analytics.AnalyticsEvent.S_SELECT_FEE
import io.muun.apollo.domain.selector.BitcoinUnitSelector
import io.muun.apollo.presentation.ui.base.SingleFragmentPresenter
import io.muun.apollo.presentation.ui.base.di.PerFragment
import newop.EditFeeState
import javax.inject.Inject

@PerFragment
class RecommendedFeePresenter @Inject constructor(
    private val bitcoinUnitSel: BitcoinUnitSelector,
) : SingleFragmentPresenter<RecommendedFeeView, RecommendedFeeParentPresenter>() {

    override fun setUp(arguments: Bundle) {
        super.setUp(arguments)
        view.setBitcoinUnit(bitcoinUnitSel.get())

        parentPresenter
            .watchEditFeeState()
            .doOnNext(view::setState)
            .doOnNext(this::reportScreen)
            .let(this::subscribeTo)
    }

    fun confirmFee(selectedFeeRateInVBytes: Double) {
        parentPresenter.confirmFee(selectedFeeRateInVBytes)
    }

    fun editFeeManually() {
        parentPresenter.editFeeManually()
    }

    fun reportScreen(state: EditFeeState) {
        // TODO this is a bit duplicated in view::setState, we could refactor a bit better
        val feeWindow = state.resolved.paymentContext.feeWindow
        val feeRateFastInVBytes = state.minFeeRateForTarget(feeWindow.fastConfTarget)
        val feeRateMidInVBytes = state.minFeeRateForTarget(feeWindow.mediumConfTarget)
        val feeRateSlowInVBytes = state.minFeeRateForTarget(feeWindow.slowConfTarget)

        analytics.report(S_SELECT_FEE(feeRateFastInVBytes, feeRateMidInVBytes, feeRateSlowInVBytes))
    }

    fun reportShowSelectFeeInfo() {
        analytics.report(S_MORE_INFO(S_MORE_INFO_TYPE.SELECT_FEE))
    }
}